### PR TITLE
[DO NOT MERGE] Enable vectorized reader #536

### PR DESCRIPTION
**Pull request by LuciferYang**
## What changes were proposed in this pull request?
1、add Option[VectorizedContext] to DataFile Constructor which contain vectorized read infomation(partitionColumns\partitionValues\returningBatch)  
2、add SpecificOapRecordReaderBase\VectorizedOapRecordReader\IndexedVectorizedOapRecordReader to implement parquet data file vectorized read, include
indexed read & full table scan.
3、copy ColumnarBatch.java from spark and add a method named "boolean isFiltered(int rowId)" provide FileSourceScanExec.doProduceVectorized method to know which rowid should be write to outbuffer when SupportBatchReturn is true
4、extract static factory method  named "IndexedParquetMetadata.from" to produce IndexedParquetMetadata with parquet footer & index rowIdList
5、add FilterHelper wap parquet filter push down logic and use pushed filter when full table scan

Corresponding doc: #536 

## How was this patch tested?
1、exiting test case

next step I will try to optimize ColumnarBatch.filteredRows markup process & add more new case to  test this issue.

### CODE REVIEW

File: src/main/java/org/apache/parquet/hadoop/InternalOapRecordReader.java [hunk 1]
```diff
@@ -116,7 +117,7 @@ public void initialize(ParquetFileReader parquetFileReader, Configuration config
         this.fileSchema = parquetFileMetadata.getSchema();
         Map<String, String> fileMetadata = parquetFileMetadata.getKeyValueMetaData();
         ReadSupport.ReadContext readContext = readSupport.init(new InitContext(
-                configuration, toSetMultiMap(fileMetadata), fileSchema));
+                configuration, Collections3.toSetMultiMap(fileMetadata), fileSchema));
```
Comment (xuanyuanking) [inline src/main/java/org/apache/parquet/hadoop/InternalOapRecordReader.java]: Why need the changing here?

File: src/main/java/org/apache/parquet/hadoop/utils/Collections3.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Jdk has Collections, Guava has Colletions2,
+ * We use Collections3.
+ */
+public final class Collections3 {
```
Comment (xuanyuanking) [inline src/main/java/org/apache/parquet/hadoop/utils/Collections3.java]: Why need this Collections3 but not keep same with parquet-hadoop

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            currentPageNumber++;
+            return true;
+        }
+    }
+
+    private void collectRowIdsToPages() {
+        this.idsMap.clear();
+        this.currentPageNumber = 0;
+        int pageSize = columnarBatch.capacity();
+        IntList currentIndexList = rowIdsIter.next();
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Admittedly, there is a one-to-one correspondence between a row group and a local row ids List and `rowIdsIter.next()` here never throws an exception as long as there is a row group, but for "Iterator Style", please add
`if (!rowIdsIter.hasNext()) return;` check before `rowIdsIter.next()` , although it's sort of redundant.

File: src/main/java/org/apache/parquet/hadoop/InternalOapRecordReader.java (hunk 1 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/InternalOapRecordReader.java]: Collection3.toSetMultiMap is duplicate in InternalOapRecordReader & SpecificOapRecordReaderBase.

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    /*
+     * Can be called before any rows are returned to enable returning columnar batches directly.
+     */
+    public void enableReturningBatches() {
+        returnColumnarBatch = true;
+    }
+
+    /**
+     * Advances to the next batch of rows. Returns false if there are no more.
+     * From VectorizedParquetRecordReader, chanage private -> protected.
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: the comment here `change private -> protected`?

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+     * From VectorizedParquetRecordReader, change private to protected
+     */
+    protected int numBatched = 0;
+
+    /**
+     * For each request column, the reader to read this column. This is NULL if this column
+     * is missing from the file, in which case we populate the attribute with NULL.
+     * From VectorizedParquetRecordReader, change private to protected, wrapper VectorizedColumnReader.
+     */
+    protected VectorizedColumnReaderWrapper[] columnReaders;
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: So what's the benefit of VectorizedColumnReaderWrapper compared to VectorizedColumnReader?

File: src/main/java/org/apache/parquet/hadoop/metadata/IndexedParquetMetadata.java [hunk 6]
```diff
@@ -17,4 +36,34 @@ public IndexedParquetMetadata(FileMetaData fileMetaData, List<BlockMetaData> blo
...
+        int nextRowGroupStartRowId = 0;
+        int totalCount = globalRowIds.length;
+        int index = 0;
+
+        for (BlockMetaData block : footer.getBlocks()) {
+            int currentRowGroupStartRowId = nextRowGroupStartRowId;
+            nextRowGroupStartRowId += block.getRowCount();
+            IntList rowIdList = new IntArrayList();
+            while (index < totalCount) {
+                int globalRowGroupId = globalRowIds[index];
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/parquet/hadoop/metadata/IndexedParquetMetadata.java]: the variable `globalRowGroupId ` refers to the global row id, right?
but the name is confusing, so rename it to `globalRowId`

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+import org.apache.spark.memory.MemoryMode;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.execution.datasources.parquet.VectorizedColumnReader;
+import org.apache.spark.sql.execution.datasources.parquet.VectorizedColumnReaderWrapper;
+import org.apache.spark.sql.execution.vectorized.ColumnVectorUtils;
+import org.apache.spark.sql.execution.vectorized.ColumnarBatch;
+import org.apache.spark.sql.types.StructField;
+import org.apache.spark.sql.types.StructType;
+
+public class VectorizedOapRecordReader extends SpecificOapRecordReaderBase<Object> {
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: diff with VectorizedParquetRecordReader,  VectorizedOapRecordReader only redefined fileds and checkEndOfRowGroup method access modifier from private to protected.

File: src/main/java/org/apache/parquet/hadoop/metadata/IndexedParquetMetadata.java [hunk 8]
```diff
@@ -17,4 +36,34 @@ public IndexedParquetMetadata(FileMetaData fileMetaData, List<BlockMetaData> blo
     public List<IntList> getRowIdsList() {
         return rowIdsList;
     }
+
+    public static IndexedParquetMetadata from(ParquetMetadata footer, int[] globalRowIds) {
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/metadata/IndexedParquetMetadata.java]: This factor method use by OapRecordReader & IndexedVectorizedOapRecordReader.

File: src/main/java/org/apache/parquet/hadoop/utils/Collections3.java (hunk 2 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/utils/Collections3.java]: Collection3.toSetMultiMap is duplicate in InternalOapRecordReader & SpecificOapRecordReaderBase.

File: src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+   * Resets the batch for writing.
+   */
+  public void reset() {
+    for (int i = 0; i < numCols(); ++i) {
+      columns[i].reset();
+    }
+    if (this.numRowsFiltered > 0) {
+      Arrays.fill(filteredRows, false);
+    }
+    // For IndexedVectorizedOapRecordReader
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: The state of current batch rows depends on last row filtering behavior, which is a little bit inappropriate.
remove  `numRowsValid` field, move this check&assign code snippet to `IndexedVectorizedOapRecordReader.filterRowsWithIndex()`

File: src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReaderWrapper.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+public class VectorizedColumnReaderWrapper {
+
+    private VectorizedColumnReader reader;
+
+    public VectorizedColumnReaderWrapper(VectorizedColumnReader reader) {
+        this.reader = reader;
+    }
+
+    public void readBatch(int total, ColumnVector column) throws IOException {
```
Comment (LuciferYang) [inline src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReaderWrapper.java]: VectorizedColumnReader.readBatch is package access modifier.

File: src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+ * The layout is a columnar with values encoded in their native format. Each RowBatch contains
+ * a horizontal partitioning of the data, split into columns.
+ *
+ * The ColumnarBatch supports either on heap or offheap modes with (mostly) the identical API.
+ *
+ * TODO:
+ *  - There are many TODOs for the existing APIs. They should throw a not implemented exception.
+ *  - Compaction: The batch and columns should be able to compact based on a selection vector.
+ */
+public final class ColumnarBatch {
```
Comment (LuciferYang) [inline src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: 1.add filed numRowsValid record valid rows count
2.add method "boolean isFiltered(int rowId) "  use by DataSourceScanExec to filter rows
3.modify "void reset()" method to reset filteredRows status by numRowsValid
4.add "void markValid(int rowId)" method use by IndexedVectorizedOapRecordReader to mark which row we need write to UnsafeRowWriter
5.add ”void markAllFiltered() “ method use by IndexedVectorizedOapRecordReader set init status

**Comment by LuciferYang** [comment]
@Moonlit-Sailor Thx~ Now IndexedVectorizedOapRecordReader#filterRowsWithIndex remove current ids and do "rowsReturned = totalCountLoadedSoFar" when idMap is empty, if "
rowsReturned == totalCountLoadedSoFar" will do readNextRowGroup

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    public boolean nextBatch() throws IOException {
+        // if idsMap is Empty, needn't read remaining data in this row group
+        // rowsReturned = totalCountLoadedSoFar to skip remaining data
+        if (idsMap.isEmpty()) {
+            rowsReturned = totalCountLoadedSoFar;
+        }
+        return super.nextBatch() && filterRowsWithIndex();
+    }
+
+    protected void checkEndOfRowGroup() throws IOException {
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: ok

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        super.initBatch(partitionColumns, partitionValues);
+        columnarBatch.markAllFiltered();
+    }
+
+    private boolean filterRowsWithIndex() throws IOException {
+        IntList ids = idsMap.remove(currentPageNumber);
+        if (ids == null || ids.isEmpty()) {
+            currentPageNumber++;
+            return this.nextBatch();
+        } else {
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Thx~  good idea~

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+            for (Integer rowid : ids) {
+                columnarBatch.markValid(rowid);
+            }
+            currentPageNumber++;
+            return true;
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: divideRowIdsIntoPages~

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+    private void collectRowIdsToPages() {
+        this.idsMap.clear();
```
[the remark this answers is not in the document]
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Perhaps Should throw a IllegalStateException, when we call divideRowIdsIntoPages method and idsMap not empty, it must be a mistake....

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java (hunk 3 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Same as idsMap, rowIdsIter must hasNext，I will add add a assert to ensure  rowIdsIter.hasNext is true

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java (hunk 5 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: VectorizedColumnReader.readBatch is package access, wrap it to public.

File: src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java (hunk 4 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/VectorizedOapRecordReader.java]: Wrong comments, nextBatch method no change.

File: src/main/java/org/apache/parquet/hadoop/metadata/IndexedParquetMetadata.java (hunk 6 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/metadata/IndexedParquetMetadata.java]: Thx~ This is a semantic error~

File: src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java (hunk 9 above)
Comment (LuciferYang) [reply src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: Already remove numRowsValid field~

File: src/main/java/org/apache/parquet/hadoop/api/VectorizedRecordReader.java [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.parquet.hadoop.api;
+
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.StructType;
+
+public interface VectorizedRecordReader<T> extends RecordReader<T> {
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/parquet/hadoop/api/VectorizedRecordReader.java]: well, I think class `VectorizedRecordReader` is unnecessary, the current inheritance hierarchy is:
　　RecordReader
　　　　　｜
　VectorizedRecordReader
　　　　　｜
SpecificOapRecordReaderBase
　　　　　｜
VectorizedOapRecordReader 
　　　　　｜
IndexedVectorizedOapRecordReader 

It is a little bit bloated, maybe we could remove the `VectorizedRecordReader` class, and make `SpecificOapRecordReaderBase` inherit `RecordReader` directly.
`SpecificOapRecordReaderBase` is for customized record reader. It seems that making `enableReturningBatches()` owned by `VectorizedOapRecordReader ` is more sensible.

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    protected void checkEndOfRowGroup() throws IOException {
+        if (rowsReturned != totalCountLoadedSoFar) return;
+        // if rowsReturned == totalCountLoadedSoFar
+        // readNextRowGroup & divideRowIdsIntoPages
+        super.readNextRowGroup();
+        this.divideRowIdsIntoPages();
+    }
+
+    @Override
+    public void initBatch(StructType partitionColumns, InternalRow partitionValues) {
```
Comment (Moonlit-Sailor) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: Is this `initBatch()` method necessary now?

Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: OK, it's unnecessary now~

File: src/main/java/org/apache/parquet/hadoop/api/VectorizedRecordReader.java (hunk 16 above)
Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/api/VectorizedRecordReader.java]: Now SpecificOapRecordReaderBase inherit RecordReader directly.

**Comment by Moonlit-Sailor** [comment]
Looks Good To Me
Thanks @LuciferYang   👍 

We would consider merging this after 0.3.0 release, during this time you might add some unit test cases for parquet vectorized reader or some comments for key functions in `IndexedVectorizedOapRecordReader` .

File: src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 18]
```diff
@@ -23,7 +23,7 @@ import org.apache.commons.lang3.StringUtils
 import org.apache.hadoop.fs.{BlockLocation, FileStatus, LocatedFileStatus, Path}
 
 import org.apache.spark.rdd.RDD
-import org.apache.spark.sql.{AnalysisException, SparkSession}
```
Comment (linhong-intel) [inline src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: We'd better keep this identical with spark.

**Comment by LuciferYang** [comment]
@linhong-intel  optimized import content by intellij :( , fix ix....

File: src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala [hunk 19]
```diff
@@ -0,0 +1,827 @@
...
+      rowEquals(it3.next(), Row(4, 4.4, 4, "world"))
+      assert(!it3.hasNext)
+
+      batch.markFiltered(2)
+      assert(batch.numValidRows() == 1)
+      val it4 = batch.rowIterator()
+      rowEquals(it4.next(), Row(null, 2.2, 2, "abc"))
+
+      // mark all filtered and verify
+      batch.markAllFiltered()
```
Comment (LuciferYang) [inline src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchSuite.scala]: test new method: markAllFiltered\markValid\isFiltered and keep numValidRows right

File: src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java [hunk 20]
```diff
@@ -0,0 +1,162 @@
...
+        for (BlockMetaData block : indexedFooter.getBlocks()) {
+            this.totalRowCount += block.getRowCount();
+        }
+        super.initializeInternal();
+    }
+
+    @Override
+    public Object getCurrentValue() throws IOException, InterruptedException {
+        if (returnColumnarBatch) return columnarBatch;
+        Preconditions.checkNotNull(batchIds, "returnColumnarBatch = false, batchIds must not null.");
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/IndexedVectorizedOapRecordReader.java]: will used when returnColumnarBatch is false

File: pom.xml [hunk 21]
```diff
@@ -303,6 +303,12 @@
       <artifactId>RoaringBitmap</artifactId>
       <version>0.6.51</version>
     </dependency>
+    <dependency>
```
Comment (LuciferYang) [inline pom.xml]: for mock object in test case

File: src/main/java/org/apache/parquet/hadoop/DefaultRecordReader.java [hunk 22]
```diff
@@ -70,6 +71,7 @@ public void initialize() throws IOException, InterruptedException {
             footer = readFooter(configuration, file, NO_FILTER);
         }
         ParquetFileReader parquetFileReader = ParquetFileReader.open(configuration, file, footer);
+        parquetFileReader.filterRowGroups(getFilter(configuration));
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/DefaultRecordReader.java]: apply data filters, copy from ParquetRecordReader, filter unnecessary rowgroup by pushed filters.

File: src/main/java/org/apache/parquet/hadoop/RecordReaderBuilder.java [hunk 23]
```diff
@@ -1,60 +0,0 @@
...
-import org.apache.parquet.hadoop.api.ReadSupport;
-import org.apache.parquet.hadoop.api.RecordReader;
-import org.apache.parquet.hadoop.metadata.ParquetMetadata;
-
-import java.io.IOException;
-
-import static org.apache.parquet.Preconditions.checkNotNull;
-
-
-public class RecordReaderBuilder<T> {
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/RecordReaderBuilder.java]: constructor parameters  not too much, remove this builder

File: src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.parquet.hadoop.metadata.BlockMetaData;
+import org.apache.parquet.hadoop.metadata.ParquetMetadata;
+import org.apache.parquet.hadoop.utils.Collections3;
+import org.apache.parquet.schema.MessageType;
+import org.apache.spark.sql.execution.datasources.oap.io.OapReadSupportImpl;
+import org.apache.spark.sql.execution.datasources.parquet.ParquetReadSupportHelper;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.types.StructType$;
+
+public abstract class SpecificOapRecordReaderBase<T> implements RecordReader<T> {
```
Comment (LuciferYang) [inline src/main/java/org/apache/parquet/hadoop/SpecificOapRecordReaderBase.java]: imitate SpecificParquetRecordReaderBase

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 25]
```diff
@@ -125,8 +125,18 @@ private[sql] class OapFileFormat extends FileFormat
    * Returns whether the reader will return the rows as batch or not.
    */
   override def supportBatch(sparkSession: SparkSession, schema: StructType): Boolean = {
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: copy form parquetfileformat , add readerClassName.equals(OapFileFormat.PARQUET_DATA_FILE_CLASSNAME) condition

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 26]
```diff
@@ -279,6 +289,19 @@ private[sql] class OapFileFormat extends FileFormat
...
+        // If true, enable using the custom RecordReader for parquet. This only works for
+        // a subset of the types (no complex types).
+        val resultSchema = StructType(partitionSchema.fields ++ requiredSchema.fields)
+        val enableVectorizedReader: Boolean =
+          m.dataReaderClassName.equals(OapFileFormat.PARQUET_DATA_FILE_CLASSNAME) &&
+          sparkSession.sessionState.conf.parquetVectorizedReaderEnabled &&
+            resultSchema.forall(_.dataType.isInstanceOf[AtomicType])
+        // Whole stage codegen (PhysicalRDD) is able to deal with batches directly
+        val returningBatch = supportBatch(sparkSession, resultSchema)
+        val pushed = FilterHelper.tryToPushFilters(sparkSession, requiredSchema, filters)
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: copy from parquetfileformat, push down filters to skip some rowgroup for batch read(not index read).

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala [hunk 27]
```diff
@@ -279,6 +289,19 @@ private[sql] class OapFileFormat extends FileFormat
         val broadcastedHadoopConf =
           sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
 
+
+        // TODO: if you move this into the closure it reverts to the default values.
+        // If true, enable using the custom RecordReader for parquet. This only works for
+        // a subset of the types (no complex types).
+        val resultSchema = StructType(partitionSchema.fields ++ requiredSchema.fields)
+        val enableVectorizedReader: Boolean =
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/OapFileFormat.scala]: copy from parquetfileformat

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/FilterHelper.scala [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.filter2.predicate.{FilterApi, FilterPredicate}
+import org.apache.parquet.hadoop.ParquetInputFormat
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.datasources.parquet.ParquetFilters
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.types.StructType
+
+object FilterHelper {
```
Comment (LuciferYang) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/io/FilterHelper.scala]: copy from parquetfileformat, extract code to FilterHelper.

**Comment by linhong-intel** [comment]
tried a simple test
the performance is as fast as native parquet reader

**Comment by LuciferYang** [comment]
@linhong-intel as which parquet reader? VectorizedParquetRecordReader ?

**Comment by linhong-intel** [comment]
Thanks @LuciferYang to split this PR into small ones.
So this is used to understand the overall design. We'll merge the small PRs.

**Comment by LuciferYang** [comment]
@linhong-intel yeah~ this is a panorama ~ as #536 said, split this pr into 7 small ones~

File: src/main/java/org/apache/parquet/hadoop/DefaultRecordReader.java [hunk 29]
```diff
@@ -70,7 +71,8 @@ public void initialize() throws IOException, InterruptedException {
             footer = readFooter(configuration, file, NO_FILTER);
         }
         ParquetFileReader parquetFileReader = ParquetFileReader.open(configuration, file, footer);
-        this.internalReader = new InternalParquetRecordReader<T>(readSupport);
+        parquetFileReader.filterRowGroups(getFilter(configuration));
```
Comment (linhong-intel) [inline src/main/java/org/apache/parquet/hadoop/DefaultRecordReader.java]: I suggest put this line to PR #576

Comment (LuciferYang) [reply src/main/java/org/apache/parquet/hadoop/DefaultRecordReader.java]: ok~ i'll add this line to #576 and remove it from #575

**Comment by LuciferYang** [comment]
All Code Already merged into master~

**Pull request closed without merging**